Add serverless, remove no longer working Trac, BuildbotTryNotifier.

danuker opened the pull request:
Closes #19 

This is a first try at the `serverless` library. I don't know where the output of the logger is, when I run it with `sls offline`.

To be continued in #25.

To run manually:

1. Install serverless NPM dependencies: `npm install`
2. Run the off-cloud version on localhost: `sls offline`

danuker commented on .npmignore:
```diff
@@ -0,0 +1,103 @@
+# Logs
```
This file was generated; no need to review.

danuker commented on chevah/github_hooks_server/handler.py:
```diff
@@ -8,8 +8,6 @@
 
 
 from chevah.github_hooks_server import log
-from chevah.github_hooks_server.utils.trac import Trac, TracException
```
I removed all Trac handling.

danuker commented on chevah/github_hooks_server/server.py:
```diff
@@ -48,47 +45,44 @@ def __init__(self, message):
+    return func.HttpResponse('pong')
 
 
-@route('/ping',  methods=['GET'])
```
The route definitions are now in `serverless.yml`.

adiroiban replied:
that is not nice ... but what can we do :)   this is how servless works

danuker commented on chevah/github_hooks_server/server.py:
```diff
@@ -114,32 +108,24 @@ def hook(request, hook_name):
-    """
-    Buildbot integration entry point.
-    """
-    return BuildbotTryNotifier(CONFIGURATION)
```
I removed `BuildbotTryNotifier` as well, since it wouldn't work after the Py3 upgrade, and I believe it is no longer used either.

adiroiban replied:
yes... buildmaster is no longer used.

danuker commented on config.toml:
```diff
@@ -3,12 +3,12 @@
+        "140.82.112.0/20",
+        "143.55.64.0/20",
+        "2a0a:a440::/29",
+        "2606:50c0::/32"
```
New GitHub URLs were in https://api.github.com/meta. But it is not used right now. Should I remove it completely?

adiroiban replied:
We  can remove... btw. we can also remove toml dependency.

This was an experiment to try toml ... but I think that for technical usage yaml is more popular and for simple things .INI is better 

so, for this project I think that is best to just use .INI

danuker commented on serverless.yml:
```diff
@@ -0,0 +1,157 @@
+# For full config options, check the docs:
+#    docs.serverless.com
+#
+# Happy Coding!
```
I kept this file close to the [original template file](https://github.com/serverless/serverless-azure-functions#create-a-new-azure-function-app) generated using `sls create`. Should I clean it up instead?

adiroiban replied:
I think that you can clean it and add a link to the docs for the format of the file.

danuker commented on setup.py:
```diff
@@ -41,31 +41,23 @@ def run(self):
-        'chardet',
-        'certifi',
-        'idna',
+        'github3.py',
```
I did not tesst the actual GitHub integration. It needs the secret, and to be deploymed.

adiroiban replied:
We can test it in a separate PR... it's ok to reduce the scope of this PR and just focus and nodejs tooling and deployment

we can fix the actual GH code in a separate PR.

the tests will need refactoring 

the private file is where you store a github API token... i think, it was a local file to prevent leaking the code via the env or process arguments.

there are some end-to-end tests and they are more suited for manual usage and not automated tests

danuker commented on serverless.yml:
```diff
@@ -0,0 +1,157 @@
+  # type: premium  # premium azure functions
+
+  environment: # these will be created as application settings
+    VARIABLE_FOO: 'foo'
```
The environment variables might need updating.

adiroiban replied:
that is ok

danuker commented:
@adiroiban I doubt this is ready for merge, but would appreciate a first look (I added comments on the diff).

adiroiban commented on setup.py:
```diff
@@ -41,31 +41,23 @@ def run(self):
+        'azure-functions-worker',
         ],
-    extras_require = {
+    extras_require={
```
don't worry too much about formating... we will move this to `setup.cfg` or `pyproject.toml` 

we might want to experiment with https://python-poetry.org/

danuker replied:
I added a tentative `pyproject.toml`. You can use it with `poetry shell`, then run what you need.

adiroiban replied:
have you pushed it ? I don't see any pyproject.toml  ... and I think that once you use pyproject.toml you can remove the setup.py  ... or have just `setup()` inside setup.py to have `pip install =e .`

Now that I am looking at poetry I see that it is sponsored by vercel and that vercel has python serverless functions

https://vercel.com/docs/concepts/functions/supported-languages#supported-languages

so maybe we can deploy this to vercel instead of azure :p

danuker replied:
No, I commented before pushing; sorry. I was experimenting with the venv also; I have pushed now.

I think I would also prefer not to use this library with tons of insecure dependencies. When pushing I got:
```
remote: GitHub found 5 vulnerabilities on chevah/github-hooks-server's default branch (2 critical, 1 high, 2 moderate). To find out more, visit:
remote:      https://github.com/chevah/github-hooks-server/security/dependabot
remote: 
```

But the pros of serverless is we can easily change to [cheaper providers](https://www.serverless.com/framework/docs/providers). Looking at the [Azure intro](https://www.serverless.com/framework/docs/providers/azure/guide/intro/) from there makes me think we don't need the `serverless-azure-functions` plugin.

I like whatever is developer-friendly and not too costly to adapt (to avoid vendor lock-in).

adiroiban replied:
we tried this "technology" 

for huge projects it might make sense to make it easy to switch between providers ... but for us this is not a big deal.

I went with Azure Functions as it had support for python serverless.

I now see that vercel also has support for Python

but it looks like the future is nodejs for serverless ...so in the end we might just embrace JS for all web related things

adiroiban commented on .gitignore:
```diff
@@ -13,3 +13,7 @@ virtualenv/
 twistd.log
+
+# NodeJS
+node_modules/
```
`node_modules` is some kind of `site-packages` from node.

We will see how we integrate with nodeenv ... we might want to have everything inside a single
`venv` directory  where we store the python venv and where we will also have `venv/node_modules`

danuker replied:
I tried to have everything installed in a virtualenv with `-g`, but it looks like Serverless plugins [don't work from a global install](https://forum.serverless.com/t/plugins-dont-work-when-installed-globally/5874). So we will have the `node_modules` folder if we use it.

adiroiban replied:
ok. no problem. we can live with node_modules ignored in the root... this is life :) 

but when possible we should prefer a nodeenv and -g

adiroiban commented on setup.py:
```diff
@@ -41,31 +41,23 @@ def run(self):
             'mock',
             'nose',
             'pyflakes',
             'pep8',
```
here we might want to add `nodeenv` and have a script that will call nodenv with a specific node version

danuker replied:
> and have a script that will call nodenv with a specific node version

I added `nodeenv` here, and added the "Development" section in the readme, but do you also want a script creating the environment?

adiroiban replied:
no. I think that you create the env only once. It's ok to have it only in the readme .

danuker replied:
I will look at `nodeenv` options, to see if I can move everything to one `node_modules` folder.

danuker replied:
I could get `serverless-azure-functions` to install globally in the venv, but it won't run from there.

```
Error ---------------------------------------------------
 
  Error: spawn /home/dan/Documents/Work/ProAtria/github-hooks-server/node_modules/.bin/func ENOENT
      at Process.ChildProcess._handle.onexit (node:internal/child_process:282:19)
      at onErrorNT (node:internal/child_process:475:16)
      at processTicksAndRejections (node:internal/process/task_queues:83:21)
 
     For debugging logs, run again after setting the "SLS_DEBUG=*" environment variable.
```

adiroiban commented on setup.py:
```diff
@@ -41,31 +41,23 @@ def run(self):
             ],
     },
-    test_suite = 'chevah.github_hooks_server.tests',
+    test_suite='chevah.github_hooks_server.tests',
```
this was experiment to check testing features in `setup.py`

You can remove it and look at running the tests with `py.test` framework

adiroiban reviewed (changes requested):
Thanks for the update.

It's a good start...but we need a few build updated and documentation updates.

I think that is best to use node via `nodeenv`

I think that we still need to learn about nodejs dev tools

```
virtualenv venv
. venv/bin/activate
nodeenv venv/node
pip install  .

# would be nice to have just npm install -g
# or nodeenv -r node-requirements.txt venv/node/
npm install -g serverless 
serverless plugin install -n serverless-azure-functions
```

For the future, we also need to make sure that all the npm packages that we use are  "cached" at https://github.com/orgs/chevah/packages

Now that I am looking at serverless ... I am not sure it's the best tool.

Maybe is best to just go with the standard azure tools :)
but we are trying :)

adiroiban commented on setup.py:
```diff
@@ -41,31 +41,23 @@ def run(self):
-        'certifi',
-        'idna',
+        'github3.py',
+        'azure-functions',
```
it's best to `pin` the exact version of these packages

adiroiban commented on README.rst:
````diff
@@ -14,3 +14,22 @@ It is not / and should not be published on PyPi.
+
+```sh
+npm install  # Required after changing `serverless.yml`.
+sls offline
````
I think that is best to use `npm install -g` from a nodeenv ...and use `serverless` offline. to reduce the confusion with the `sls` command

and I think that the readme should be updated to instruct that `pip install .` should be called before `servereless offline` to get the packages  and prepare the deployment environment

danuker replied:
I replaced `sls` with `serverless` everywhere. I used `poetry install` instead of `pip install .`, but `pip` should work also.

adiroiban commented on package.json:
```diff
@@ -0,0 +1,14 @@
+  },
+  "author": "",
+  "license": "ISC",
+  "devDependencies": {
```
why is this dev?
this is not needed for the deployment?

danuker replied:
In the repo it is now under `dependencies`, but I noticed something else editing it back to `devDependencies`, but I have to find out what exactly.

adiroiban commented on local.settings.json:
```diff
@@ -0,0 +1 @@
+{"IsEncrypted":false,"Values":{"AzureWebJobsStorage":"UseDevelopmentStorage=true","FUNCTIONS_WORKER_RUNTIME":"python"}}
```
is this file needed?

danuker replied:
Looks like not; I removed it.

danuker commented:
I added `poetry` support and cleaned up the docs some more.

Should I remove `setup.py` now that there is `pyproject.toml`?

I have not added the PublishCommand; there is [no command execution support for poetry](https://github.com/python-poetry/poetry/issues/265). Should I add it to Makefile?

needs-review

adiroiban commented:
For publish command is ok.


```
[[tool.poetry.source]]
name = "chevah"
default = true
url = "http://deag.chevah.com:10042"

--

poetry publish -r chevah
```


but now I found out  that while we have removed Trac...we still need those user/pass to publish pypi packages.

We should look at a replacement.

I think that this is ok for a first merge.

------------

we can then set a hook only for this repo and run some tests.

Thanks!
